api/api_signer.go

```diff
@@ -26,30 +26,51 @@ type SignHandler struct {
 // is not an empty string, the handler will send signature requests to
 // the CFSSL instance contained in remote by default.
 func NewSignHandler(caFile, caKeyFile string, policy *config.Signing) (http.Handler, error) {
-	var err error
-	s := new(SignHandler)
-
 	root := universal.Root{
-		CertFile: caFile,
-		KeyFile:  caKeyFile,
+		Config: map[string]string{
+			"cert-file": caFile,
+			"key-file":  caKeyFile,
+		},
 	}
-	if s.signer, err = universal.NewSigner(root, policy); err != nil {
+	s, err := universal.NewSigner(root, policy)
+	if err != nil {
 		log.Errorf("setting up signer failed: %v", err)
 		return nil, err
 	}
 
-	return HTTPHandler{s, "POST"}, nil
+	return NewSignHandlerFromSigner(s)
 }
 
 // NewSignHandlerFromSigner generates a new SignHandler directly from
 // an existing signer.
-func NewSignHandlerFromSigner(signer signer.Signer) HTTPHandler {
+func NewSignHandlerFromSigner(signer signer.Signer) (h HTTPHandler, err error) {
+	policy := signer.Policy()
+	if policy == nil {
+		err = errors.New(errors.PolicyError, errors.InvalidPolicy)
+		return
+	}
+
+	// Sign will only respond for profiles that have no auth provider.
+	// So if all of the profiles require authentication, we return an error.
+	haveUnauth := (policy.Default.Provider == nil)
```

Comment:
Maybe we should do this before calling NewSignHandler (and not necessarily error).